File: src/DataScience/Julia/proxy.jl
mutable struct DocumentBackend
    fetches::Int
end

function fetch!(backend::DocumentBackend, id::Int)
    backend.fetches += 1
    "doc($id)"
end

mutable struct DocumentProxy
    backend::Union{Nothing, DocumentBackend}
    cache::Dict{Int, String}
end

DocumentProxy() = DocumentProxy(nothing, Dict{Int, String}())

function get_document!(proxy::DocumentProxy, id::Int)
    get!(proxy.cache, id) do
        if isnothing(proxy.backend)
            proxy.backend = DocumentBackend(0)
        end
        fetch!(proxy.backend::DocumentBackend, id)
    end
end

proxy = DocumentProxy()
first = get_document!(proxy, 42)
second = get_document!(proxy, 42)
backend_count = isnothing(proxy.backend) ? 0 : 1
fetches = isnothing(proxy.backend) ? 0 : (proxy.backend::DocumentBackend).fetches
println("backend=$backend_count;fetches=$fetches;first=$first;second=$second")

File: src/DataScience/MATLAB/proxy.m
backendCreated = 0;
fetches = 0;
cache = containers.Map('KeyType', 'double', 'ValueType', 'char');

[first, backendCreated, fetches, cache] = proxyGet(42, backendCreated, fetches, cache);
[second, backendCreated, fetches, cache] = proxyGet(42, backendCreated, fetches, cache);

assert(backendCreated == 1, 'Proxy must lazily create exactly one backend.');
assert(fetches == 1, 'Second read must be served from cache.');
fprintf('backend=%d;fetches=%d;first=%s;second=%s\n', backendCreated, fetches, first, second);

function [value, backendCreated, fetches, cache] = proxyGet(id, backendCreated, fetches, cache)
    if isKey(cache, id)
        value = cache(id);
        return;
    end

    if backendCreated == 0
        backendCreated = 1;
    end

    fetches = fetches + 1;
    value = sprintf('doc(%d)', id);
    cache(id) = value;
end
